c/src/reduce.cu

```diff
@@ -0,0 +1,866 @@
+      case cccl_type_enum::INT8:
+        check(nvrtcGetTypeName<::cuda::std::int8_t>(&result));
+        break;
+      case cccl_type_enum::INT16:
+        check(nvrtcGetTypeName<::cuda::std::int16_t>(&result));
+        break;
+      case cccl_type_enum::INT32:
+        check(nvrtcGetTypeName<::cuda::std::int32_t>(&result));
+        break;
+      case cccl_type_enum::INT64:
+        check(nvrtcGetTypeName<::cuda::std::int64_t>(&result));
+        break;
+      case cccl_type_enum::UINT8:
+        check(nvrtcGetTypeName<::cuda::std::uint8_t>(&result));
+        break;
+      case cccl_type_enum::UINT16:
+        check(nvrtcGetTypeName<::cuda::std::uint16_t>(&result));
+        break;
+      case cccl_type_enum::UINT32:
+        check(nvrtcGetTypeName<::cuda::std::uint32_t>(&result));
+        break;
+      case cccl_type_enum::UINT64:
+        check(nvrtcGetTypeName<::cuda::std::uint64_t>(&result));
+        break;
+      case cccl_type_enum::FLOAT32:
+        check(nvrtcGetTypeName<float>(&result));
+        break;
+      case cccl_type_enum::FLOAT64:
+        check(nvrtcGetTypeName<double>(&result));
+        break;
+      case cccl_type_enum::STORAGE:
+        check(nvrtcGetTypeName<storage_t>(&result));
+        break;
+    }
+  }
+
+  return result;
+}
+
+struct reduce_tuning_t
+{
+  int cc;
+  int block_size;
+  int items_per_thread;
+  int vector_load_length;
+};
+
+template <int N>
+reduce_tuning_t find_tuning(int cc, const reduce_tuning_t (&tunings)[N])
+{
+  for (const reduce_tuning_t& tuning : tunings)
+  {
+    if (cc >= tuning.cc)
+    {
+      return tuning;
+    }
+  }
+
+  return tunings[N - 1];
+}
+
+runtime_tuning_policy get_policy(int cc, cccl_type_info accumulator_type, cccl_type_info input_type)
+{
+  reduce_tuning_t chain[] = {{60, 256, 16, 4}, {35, 256, 20, 4}};
+
+  auto [_, block_size, items_per_thread, vector_load_length] = find_tuning(cc, chain);
+
+  // Implement part of MemBoundScaling
+  items_per_thread = CUB_MAX(1, CUB_MIN(items_per_thread * 4 / accumulator_type.size, items_per_thread * 2));
+  block_size       = CUB_MIN(block_size, (((1024 * 48) / (accumulator_type.size * items_per_thread)) + 31) / 32 * 32);
+
+  return {block_size, items_per_thread, vector_load_length};
+}
+
+cccl_type_info get_accumulator_type(cccl_op_t op, cccl_iterator_t input_it, cccl_value_t init)
+{
+  // TODO Should be decltype(op(init, *input_it)) but haven't implemented type arithmetic yet
+  //      so switching back to the old accumulator type logic for now
+  return init.type;
+}
+
+cudaError_t InvokeSingleTile(
+  void* d_temp_storage,
+  std::size_t& temp_storage_bytes,
+  cccl_iterator_t d_in,
+  cccl_iterator_t d_out,
+  unsigned long long num_items,
+  cccl_op_t op,
+  cccl_value_t init,
+  int cc,
+  CUfunction single_tile_kernel,
+  CUstream stream)
+{
+  const runtime_tuning_policy policy = get_policy(cc, d_in.value_type, d_in.value_type);
+
+  cudaError error = cudaSuccess;
+  do
+  {
+    if (d_temp_storage == NULL)
```

Comment:
```suggestion
    if (d_temp_storage == nullptr)
```

Reply:
This brings up the question what C version we are actually targetting. `nullptr` was introduced in C23, which is super new. It's also a complete mockery on us C++11 devs in the rest of the library :D

Reply:
🤔 this is not decorated with extern C, so I would assume its a C++ function

Reply:
Oh, I didn't check the compilation language and assumed it was C. In any case, if we have any pure C sources, we should settle on the C standard used.